### ISSUE TITLE
Court User: Linking of two or more "like" cases

### ISSUE THREAD

**Issue by mariahkannenberg** [flexion/ef-cms#482]
As a court user, so that I may easily manage two or more like cases, I need the ability to link the cases in the system.

## Acceptance Criteria
* Court user can link two or more like cases in the system
   * Must have the same case procedure (Small or Regular)
   * Must have the same requested place of trial
   * Must have the same Status
   * Must have the same Judge
* Court user can easily see which cases are linked in the system
* In order to be linked, cases must be (at least) General Docket - At Issue
* Cases can't be consolidated if they're in the following statuses:
   * New
   * Batched for IRS
   * Recalled
   * General Docket - Not at Issue
   * Closed (Note: if a case is consolidated before going into this status, it may stay consolidated while in this status; it may not be set to consolidated while in this status)
   * On Appeal (Note: if a case is consolidated before going into this status, it may stay consolidated while in this status; it may not be set to consolidated while in this status)

## Security Considerations
* Does this work make you nervous about privacy or security?
* Does this work make major changes to the system?
* Does this work implement new authentication or security controls?
* Does this work create new methods of authentication, modify existing security controls, or explicitly implement any security or privacy features?

## Notes
* Doesn't include submission of documents
Notes from 12/4 Design Meeting:
* Order to Consolidate - any automation of this related to actual consolidation is not part of MVP or this story.
* Consolidation card displays on every case detail page. However, if case is not in a state that allows consolidation, show a message in the card
* Cases don't need to be the same in order to be consolidated
* Unlinking is not part of this story
* Dialogue popup for consolidated icon should be post-MVP
* Show/hide nesting feature - remove - not part of MVP
* Remove consolidation icon from all places except the case detail page, document detail page, and external users dashboard. 

## Definition of Done (Updated 8-28-19)
**Product Owner**
 - [ ]  Acceptance criteria have been met

**UX**
 - [x] Business test scenarios to meet all acceptance criteria have been written
 - [ ] Usability has been validated
 - [x] Wiki has been updated (if applicable) 
 - [x] Story has been tested on a mobile device (for external users only)

**Engineering**
- [x] Automated test scripts have been written (JoJo)
- [x] Field level and page level validation errors (front-end and server-side) integrated and functioning (Matt)
 - [x] New screens have been added to pa11y scripts
 - [x] All new functionality verified to work with keyboard and macOS voiceover https://www.apple.com/voiceover/info/guide/_1124.html (Matt #3833)
 - [x] READMEs, other appropriate docs, JSDocs and swagger/APIs fully updated (Matt #3833)
 - [x] UI should be touch optimized and responsive for external only (functions on supported mobile devices and optimized for screen sizes as required)
 - [x] Module dependencies are up-to-date and are at the latest resolvable version (npm update) (Jojo)
 - [x] Errors in Sonarcloud are fixed https://sonarcloud.io/organizations/flexion-github/projects (Jojo)
 - [x] Lambdas include CloudWatch logging of users, inputs and outputs
 - [x] Code refactored for clarity and to remove any known technical debt 
 - [x] Deployed to the dev environment
 - [x] Deployed to the stage environment

**Comment by alsmith1** [comment]
From notes with Rhonda: Certain doc types can’t be filed in more than one case - needs to be rule based 
Ex: Entry of Appearance
Amended Petition
Amended Answer

**Comment by alsmith1** [comment]
## Test scenarios
#### Scenario 1 - If case is consolidated It will display a yellow 2 paper icon to the left of the Docket Number in the header
1. Login as any internal/ external user
2.  Find a Case that has been consolidated
3. Verify that the header has a yellow 2 paper icon to the left of the Docket Number in the header

Expected result
- [ ] Header displays a yellow 2 paper icon to the left of the Docket Number in the header

#### Scenario 2 - Court user can easily see which cases are linked in the system (Messages)
1. Login as any internal user
2. Navigate to the My Messages Inbox page
3. Verify that any consolidated cases have a yellow 2 paper icon to the left of the Docket Number
4. Navigate to the My Messages Sent page
5. Verify that any consolidated cases have a yellow 2 paper icon to the left of the Docket Number
6. Navigate to the Section Messages Inbox page
7. Verify that any consolidated cases have a yellow 2 paper icon to the left of the Docket Number
8. Navigate to the Section Messages Sent page
9. Verify that any consolidated cases have a yellow 2 paper icon to the left of the Docket Number

Expected result
- [ ] Docket Number displays a yellow 2 paper icon to the left on the My Messages Inbox
- [ ] Docket Number displays a yellow 2 paper icon to the left on the My Messages Sent 
- [ ] Docket Number displays a yellow 2 paper icon to the left on the Section Messages Inbox
- [ ] Docket Number displays a yellow 2 paper icon to the left on the Section Messages Sent

#### Scenario 3 - Court user can easily see which cases are linked in the system (Document QC)
1. Login as any internal user
2. Navigate to the My Document QC Inbox page
3. Verify that any consolidated cases have a yellow 2 paper icon to the left of the Docket Number
4. Navigate to the My Document QC In Progress page
5. Verify that any consolidated cases have a yellow 2 paper icon to the left of the Docket Number
6. Navigate to the My Document QC Processed page
7. Verify that any consolidated cases have a yellow 2 paper icon to the left of the Docket Number
8. Navigate to the Section Document QC Inbox page
9. Verify that any consolidated cases have a yellow 2 paper icon to the left of the Docket Number
10. Navigate to the Section Document QC In Progress page
11. Verify that any consolidated cases have a yellow 2 paper icon to the left of the Docket Number
12. Navigate to the Section Document QC Processed page
13. Verify that any consolidated cases have a yellow 2 paper icon to the left of the Docket Number

Expected result
- [ ] Docket Number displays a yellow 2 paper icon to the left on the My Messages Inbox
- [ ] Docket Number displays a yellow 2 paper icon to the left on the My Messages In Progress 
- [ ] Docket Number displays a yellow 2 paper icon to the left on the My Messages Processed 
- [ ] Docket Number displays a yellow 2 paper icon to the left on the Section Messages Inbox
- [ ] Docket Number displays a yellow 2 paper icon to the left on the Section Messages In Progress 
- [ ] Docket Number displays a yellow 2 paper icon to the left on the Section Messages Processed 

#### Scenario 4 - Non-Court user can easily see which cases are linked in the system (Dashboard)
1. Login as an external user (petitioner/practitioner)
2. Navigate to the My Messages Inbox page
3. Verify that any consolidated cases have a yellow 2 paper icon to the left of the Docket Number
4. Navigate to the My Messages Sent page
5. Verify that any consolidated cases have a yellow 2 paper icon to the left of the Docket Number
6. Navigate to the Section Messages Inbox page
7. Verify that any consolidated cases have a yellow 2 paper icon to the left of the Docket Number
8. Navigate to the Section Messages Sent page
9. Verify that any consolidated cases have a yellow 2 paper icon to the left of the Docket Number

Expected result
- [ ] Docket Number displays a yellow 2 paper icon to the left on the My Messages Inbox
- [ ] Docket Number displays a yellow 2 paper icon to the left on the My Messages Sent 
- [ ] Docket Number displays a yellow 2 paper icon to the left on the Section Messages Inbox
- [ ] Docket Number displays a yellow 2 paper icon to the left on the Section Messages Sent

#### Scenario 5 - Court user can link two or more like cases in the system (Case Detail- Internal users)
1. Login as any internal user
2. Select a case and navigate to Docket Page > Case Info Tab
3. Verify the consolidated cases card displays with a link to add cases.
4. Click on the add cases link
5. Verify that a modal to consolidate cases displays
6. Search for a docket number
7. When result appears click the checkbox and click link cases
8. When the Case Information Tab reloads verify the consolidated cases tab displays a case

Expected result
- [ ] Consolidated cases card displays with a add cases link
- [ ] Modal Displays
- [ ] Docket entry displays after search
- [ ]  Case that was just added in modal displays in the consolidated cases tile on the case information tab

#### Scenario 6 - Court user can remove case from consolidated cases list (Case Detail- Internal users)
1. Login as any internal user
2. Select a case and navigate to Docket Page > Case Info Tab
3. Verify the consolidated cases card displays with a link to add cases and at least 1 case already added.
4. Click on the Remove Link
5. Verify modal displays to confirm removal of case from consolidated
6. When the Case Information Tab reloads verify the consolidated cases tab does not display “just removed” case

Expected result
- [ ] Consolidated cases card displays with a add cases link and at least 1 case
- [ ] Modal Displays
- [ ]  Case that was just removed in modal does not displays in the consolidated cases tile on the case information tab

#### Scenario 7 - Consolidated Cases Show docket number and case name (Case Detail- Internal users)
1. Login as any internal user
2. Select a case and navigate to Docket Page > Case Info Tab
3. Verify the consolidated cases card displays with a link to add cases and at least 1 case already added.
4. Verify the case listed under the Consolidated Cases tile displays the Docket number (link) and case name

Expected result
- [ ] Consolidated cases card displays with a add cases link and at least 1 case
- [ ] Case listed under the Consolidated Cases tile displays the Docket number (link) and case name

#### Scenario 8 - Consolidated Cases are displayed in order of docket number lowest to highest (Case Detail- all users)
1. Login as any internal user
2. Select a case and navigate to Docket Page > Case Info Tab
3. Verify the consolidated cases card displays with multiple cases listed (current case included).
4. Verify the cases are listed in order of docket number lowest to highest

Expected result
- [ ] Consolidated cases card displays with multiple case listings
- [ ] Cases listed under the Consolidated Cases display in order by docket lowest to highest
- [ ] Case in which you are currently on displays in the consolidated cases card


#### Scenario 9 - Court user can easily see which cases are linked in the system (Pending Report)
1. Login as any internal user
2. From the Reports menu select Pending Report
3. Verify that any consolidated cases has a yellow 2 paper icon to the left of the Docket Number

Expected result
- [ ] A yellow 2 paper icon to the left of the docket number for a Consolidated Case

#### Scenario 10 - Court user can easily see which cases are linked in the system (Deadlines report)
1. Login as any internal user
2. From the Reports menu select Deadlines Report
3. On the calendar select a time frame for deadlines report
4. Verify that any consolidated cases has a yellow 2 paper icon to the left of the Docket Number

Expected result
- [ ] A yellow 2 paper icon to the left of the docket number for a Consolidated Case

#### Scenario 11 - Court user can easily see which cases are linked in the system (Blocked cases report)
1. Login as any internal user
2. From the Reports menu select Blocked Cases Report
3. In the search criteria select a trial location
4. Verify that any consolidated cases has a yellow 2 paper icon to the left of the Docket Number

Expected result
- [ ] A yellow 2 paper icon to the left of the docket number for a Consolidated Case

#### Scenario 12 - Court user can easily see which cases are linked in the system (Advanced Search)
1. Login as any internal user
2. Click the Advanced search link next to the search box
3. Enter search criteria
4. Verify that any consolidated cases has a yellow 2 paper icon to the left of the Docket Number

Expected result
- [ ] A yellow 2 paper icon to the left of the docket number for a Consolidated Case


________________________________________________________________________________________________________________________________

#### Scenario 8 - To be consolidated cases Must Be (at least) General Docket - At Issue, Have the same case type (regular or small) and the same place of trial (need an order to consolidate cases - Jessica to Verify)
1. Login as any internal user
2. Select a case and navigate to Docket Page > Case Info Tab
3. Verify the consolidated cases card displays with a link to add cases and at least 1 case already added.
4. Verify the case listed under the Consolidated Cases tile displays the Docket number (link) and case name

Expected result
- [ ] Consolidated cases card displays with a add cases link and at least 1 case
- [ ] Case listed under the Consolidated Cases tile displays the Docket number (link) and case name

#### Scenario 9 - Cases can't be consolidated if they're in the following statuses: (New, Batched for IRS, Recalled, General Docket - Not at Issue, Closed (Note: if a case is consolidated before going into this status, it may stay consolidated while in this status; it may not be set to consolidated while in this status), or On Appeal (Note: if a case is consolidated before going into this status, it may stay consolidated while in this status; it may not be set to consolidated while in this status)(do not have an order to consolidate cases- Jessica to Verify)
1. Login as any internal user
2. Select a case and navigate to Docket Page > Case Info Tab
3. Verify the consolidated cases card displays with a link to add cases and at least 1 case already added.
4. Verify the case listed under the Consolidated Cases tile displays the Docket number (link) and case name

Expected result
- [ ] Consolidated cases card displays with a add cases link and at least 1 case
- [ ] Case listed under the Consolidated Cases tile displays the Docket number (link) and case name

**Comment by alsmith1** [comment]
Waiting on answer from court to see if we will be relating items as well as consolidating them- Jessica to provide next week

**Comment by alsmith1** [comment]
![Screen Shot 2019-11-26 at 12.55.39 PM.png](https://images.zenhubusercontent.com/5d766a771251c6000116623d/691f7f21-1d20-4ef7-b262-9e7bf92dc880)![Screen Shot 2019-11-26 at 12.55.46 PM.png](https://images.zenhubusercontent.com/5d766a771251c6000116623d/5539c124-6d0d-4094-83ff-78b181e998dc)![Screen Shot 2019-11-26 at 12.56.05 PM.png](https://images.zenhubusercontent.com/5d766a771251c6000116623d/e1095c6f-577d-407b-93d4-0cd676b23680)![Screen Shot 2019-11-26 at 12.56.14 PM.png](https://images.zenhubusercontent.com/5d766a771251c6000116623d/5f4b36e4-a0df-4bb3-bea5-8dd0d782de8c)![Screen Shot 2019-11-26 at 12.56.19 PM.png](https://images.zenhubusercontent.com/5d766a771251c6000116623d/11e30fba-7295-4e33-bb23-f07e6b116ba0)

**Comment by alsmith1** [comment]
![Screen Shot 2019-11-26 at 12.56.29 PM.png](https://images.zenhubusercontent.com/5d766a771251c6000116623d/d95c54c0-c217-412b-adbd-4c0090e9d512)![Screen Shot 2019-11-26 at 12.56.35 PM.png](https://images.zenhubusercontent.com/5d766a771251c6000116623d/b28f37a9-bc58-4ac5-9a3e-be4d6f373bcf)![Screen Shot 2019-11-26 at 12.56.45 PM.png](https://images.zenhubusercontent.com/5d766a771251c6000116623d/27ff2434-8aaa-4de6-9a28-23c6d6df97ef)![Screen Shot 2019-11-26 at 12.56.53 PM.png](https://images.zenhubusercontent.com/5d766a771251c6000116623d/aec49760-8a53-4049-a7e8-8e190bd51a55)

**Comment by alsmith1** [comment]
![Screen Shot 2019-11-26 at 12.57.02 PM.png](https://images.zenhubusercontent.com/5d766a771251c6000116623d/c477a509-79ce-4f9d-abf2-c35f4ddbe256)![Screen Shot 2019-11-26 at 12.57.08 PM.png](https://images.zenhubusercontent.com/5d766a771251c6000116623d/2cd3ca29-5812-4d76-b946-c711ee624801)![Screen Shot 2019-11-26 at 12.57.20 PM.png](https://images.zenhubusercontent.com/5d766a771251c6000116623d/958bf4da-1ebf-4262-aed1-d37062a8d744)![Screen Shot 2019-11-26 at 12.57.43 PM.png](https://images.zenhubusercontent.com/5d766a771251c6000116623d/4cde9f8c-11c0-47e7-b251-1d95627c95be)![Screen Shot 2019-11-26 at 12.57.58 PM.png](https://images.zenhubusercontent.com/5d766a771251c6000116623d/f42dad88-a1cc-4e40-a41e-ae9cedd84ce9)![Screen Shot 2019-11-26 at 12.58.12 PM.png](https://images.zenhubusercontent.com/5d766a771251c6000116623d/84df891e-afcb-4601-8c02-44d887310460)

**Comment by alsmith1** [comment]
![Screen Shot 2019-12-04 at 10.48.16 AM.png](https://images.zenhubusercontent.com/5d766a771251c6000116623d/8b926890-3bd5-4cc3-a84a-72c3bb2d511f)![Screen Shot 2019-12-04 at 10.48.26 AM.png](https://images.zenhubusercontent.com/5d766a771251c6000116623d/cd0a49fb-0e52-456c-9f10-d185cb9e5aa4)![Screen Shot 2019-12-04 at 10.48.33 AM.png](https://images.zenhubusercontent.com/5d766a771251c6000116623d/c80e2d27-92a5-4af8-b366-048a938774d7)![Screen Shot 2019-12-04 at 10.48.49 AM.png](https://images.zenhubusercontent.com/5d766a771251c6000116623d/1557d7d3-8ffc-4176-b161-c895113619c8)![Screen Shot 2019-12-04 at 10.48.59 AM.png](https://images.zenhubusercontent.com/5d766a771251c6000116623d/02148272-6760-42fa-af0f-26c0c971ff67)![Screen Shot 2019-12-04 at 10.49.04 AM.png](https://images.zenhubusercontent.com/5d766a771251c6000116623d/259dbb17-4601-418e-9dec-9a61080dda39)![Screen Shot 2019-12-04 at 10.49.18 AM.png](https://images.zenhubusercontent.com/5d766a771251c6000116623d/e42849e4-a7c8-4b38-b97d-2f5474d355d1)![Screen Shot 2019-12-04 at 10.49.58 AM.png](https://images.zenhubusercontent.com/5d766a771251c6000116623d/074b676e-9849-438a-8170-6508d7315593)[Uploading... Screen Shot 2019-12-04 at 10.50.16 AM.png ]![Screen Shot 2019-12-04 at 10.50.24 AM.png](https://images.zenhubusercontent.com/5d766a771251c6000116623d/78a219ca-410a-43f5-ac0c-732f5659a7ef)

**Comment by JessicaMarine** [comment]
PO Review 12.13.19:
* (1) if there is more than 1 reason that two cases cannot be consolidated, I think that all reasons should display on the modal -- right now it is only showing that the case procedure isn't the same for 121-19 and 122-19SL on staging when the Judges aren't the same and the Place of Trial isn't the same.  
* (2) if a case isn't eligible to be consolidated (like 140-19) if I try to select that case for consolidation from 138-19, the modal should probably tell me the case isn't eligible for consolidation instead of saying case status isn't the same.

**Comment by JessicaMarine** [comment]
PO Review 12/13/19 4pm:
* When you are in the modal and search for case, it is not returning any of the suffix information on the docket number.  For example, 115-19S comes up as just 115-19; 117-19W comes up as just 117-W.  That suffix really is part of the docket number and has to display
